src/main.cpp

```diff
@@ -912,16 +912,121 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+    size_t nConflictingSize = 0;
+
     // Check for conflicts with in-memory transactions
     {
     LOCK(pool.cs); // protect pool.mapNextTx
+    const CTransaction *ptxConflicting = NULL;
     for (unsigned int i = 0; i < tx.vin.size(); i++)
     {
         COutPoint outpoint = tx.vin[i].prevout;
         if (pool.mapNextTx.count(outpoint))
         {
-            // Disable replacement feature for now
-            return false;
+            if (ptxConflicting)
+            {
+                // For simplicity we only replace transactions on a 1-for-1 basis.
+                if (pool.mapNextTx[outpoint].ptx != ptxConflicting)
+                    return state.DoS(0, error("AcceptToMemoryPool: replacement %s conflicts with more than one other transaction",
+                                              hash.ToString()),
+                                     REJECT_DUPLICATE, "bad-txns-inputs-spent");
+            }
+            else
+            {
+                ptxConflicting = pool.mapNextTx[outpoint].ptx;
+                hashConflicting = ptxConflicting->GetHash();
+
+                if (GetBoolArg("-firstseensafe", true))
+                {
+                    // Check that the replacement is first-seen-safe.
+                    //
+                    // This requires the conflicting transaction's vout to be a
+                    // subset of the replacement. Outputs can pay more than before,
+                    // but not less. For simplicity we require the order of outputs
+                    // be unchanged.
+                    if (tx.vout.size() < ptxConflicting->vout.size())
+                    {
+                        return state.DoS(0, error("AcceptToMemoryPool: replacement has %s fewer outputs than original %s",
+                                                  hash.ToString(),
+                                                  hashConflicting.ToString()),
+                                         REJECT_DUPLICATE, "bad-txns-inputs-spent");
+                    }
+                    for (unsigned int j = 0; j < ptxConflicting->vout.size(); j++)
+                    {
+                        if (ptxConflicting->vout[j].scriptPubKey != tx.vout[j].scriptPubKey ||
+                            ptxConflicting->vout[j].nValue > tx.vout[j].nValue)
+                        {
+                            return state.DoS(0, error("AcceptToMemoryPool: replacement %s outputs not a subset of original %s",
+                                                      hash.ToString(),
+                                                      hashConflicting.ToString()),
+                                             REJECT_DUPLICATE, "bad-txns-inputs-spent");
+                        }
+                    }
+                }
+
+                // Make sure the outputs of the transaction we're replacing
+                // have not been spent.
+                for (unsigned int j = 0; j < ptxConflicting->vout.size(); j++)
```

Comment:
As in, check if the mapNextTx.lower_bound() hash == hashConflicting or the lower_bound() returns mapNextTx.end()?

Seems reasonable, though the way that depends on which way COutPoint's is sorted bothers me slightly.